internal/controller/streams.go

```diff
@@ -197,105 +164,225 @@ func (sm *streamManager) BootUpStreamAndReconnect(ctx context.Context, cancel co
+			randomInt, err := rand.Int(rand.Reader, max)
 			if err != nil {
-				logger.Errorw("Failed to register cluster", "error", err)
-				continue
-			}
-			err = authn.WriteK8sSecret(ctx, responseData, envMap.ClusterCreds)
-			time.Sleep(1 * time.Second)
-			if err != nil {
-				logger.Errorw("Failed to write secret to Kubernetes", "error", err)
-			}
-			clientID, clientSecret, err = authn.ReadCredentialsK8sSecrets(ctx, envMap.ClusterCreds)
-			if err != nil {
-				logger.Errorw("Could not read K8s credentials", "error", err)
+				logger.Errorw("Could not generate a random int", "error", err)
 				continue
 			}
+			result := randomInt.Int64()
+			sleep := 1*time.Second + backoff + time.Duration(result)*time.Millisecond // Add randomness
+			time.Sleep(sleep)
+			backoff = backoff * 2 // Exponential increase
+			rebootCounter++
 		}
-		conn, err := SetUpOAuthConnection(ctx, logger, envMap.TokenEndpoint, envMap.TlsSkipVerify, clientID, clientSecret)
+	}
+}
+
+// ExponentialStreamConnect will continue to reboot and restart the main operations within the operator if any disconnects or errors occur.
+func ExponentialStreamConnect(ctx context.Context, logger *zap.SugaredLogger, envMap EnvironmentConfig, bufferedGrpcSyncer *BufferedGrpcWriteSyncer) {
+	for {
+		authConn, client, err := streamAuth(ctx, logger, envMap)
 		if err != nil {
-			logger.Errorw("Failed to set up an OAuth connection", "error", err)
+			logger.Errorw("Failed to establish initial connection; will retry", "error", err)
+			time.Sleep(5 * time.Second) // Retry after a delay
 			continue
 		}
-		client, err := NewStreamClient(ctx, logger, conn)
-		if err != nil {
-			logger.Errorw("Failed to create a new stream", "error", err)
-			continue
+
+		streamClient := &streamClient{
+			conn:            authConn,
+			client:          client,
+			ciliumNamespace: envMap.CiliumNamespace,
 		}
 
 		sm := &streamManager{
-			streamClient: client,
-			logger:       logger,
+			streamClient:       streamClient,
+			logger:             logger,
+			bufferedGrpcSyncer: bufferedGrpcSyncer,
 		}
 
-		// Update the gRPC client and connection in BufferedGrpcWriteSyncer
-		bufferedGrpcSyncer.UpdateClient(client.logStream, client.conn)
-		go bufferedGrpcSyncer.ListenToLogStream()
+		resourceDone := make(chan struct{})
+		logDone := make(chan struct{})
+		ciliumDone := make(chan struct{})
+		sm.bufferedGrpcSyncer.done = logDone
+
+		go manageStream(logger, connectAndStreamResources, sm, resourceDone)
+		go manageStream(logger, connectAndStreamLogs, sm, logDone)
+
+		// Need to block this until resources are done being committed. Using waitgroups? Then we need to pass that in? Or pullout the snapschat complete watigroup to here?
+		go manageStream(logger, connectAndStreamCiliumNetworkFlows, sm, ciliumDone)
+
+		<-ciliumDone
+		<-resourceDone
+		<-logDone
+
+		logger.Warn("All streams have been closed. Rebooting the entire connection.")
```

Comment:
That logic doesn't work if we have issues with just one type of stream, as it could well be that the logs stream keeps terminating, we retry 10 times, and then `manageStream` returns for that one stream. But if the other streams were healthy, then we would never retry again to reopen the logs stream until we restart the pod.

It would make more sense to me to have `manageStream` loop indefinitely, and to cap the backoff delay, as I commented above.

You could still have `manageStream` return an error in the channels. In this case, it would now indicate a several failure, and that should cause restarting all streams.
Like:
```suggestion
		select {
		case <-ciliumDone:
		case <-resourceDone:
		case <-logDone:
		}

		logger.Warn("One or more streams have been closed; reopening the connection")
```